File: Top_100_Liked_Questions/230._Kth_Smallest_Element_in_a_BST/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        stack = []
        
        while True:
            while root:
                stack.append(root)
                root = root.left
            
            root = stack.pop()
            
            k -= 1
            if not k:
                return root.val
            
            root = root.right
            
        raise Expection("answer not found")
        
    
    def kthSmallest2(self, root: TreeNode, k: int) -> int:
        # inorder traversal
        def inorder(node):
            return inorder(node.left) + [node.val] + inorder(node.right) if node else []
        print(inorder(root))
        return inorder(root)[k-1]

File: Top_100_Liked_Questions/230._Kth_Smallest_Element_in_a_BST/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    let stack = [];
    
    while (true) {
        while (root) {
            stack.push(root);
            root = root.left;
        }
        root = stack.pop();
        
        k -= 1;
        if (k === 0) {
            return root.val;
        }
        
        root = root.right;
    }
    throw "answer not found";
}
    
var kthSmallest = function(root, k) {
    let inorder = function(node) {
        return node ? [...inorder(node.left), node.val, ...inorder(node.right)] : [];
    }
    return inorder(root)[k-1];
};